File: backend/app/schemas/graph.py
"""Compact, explicitly typed relationship explorer contracts."""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

GraphEntityType = Literal["organization", "venue", "space", "event", "event_date", "user"]
GraphRelationType = Literal[
    "organization_has_venue",
    "venue_has_space",
    "organization_has_event",
    "event_has_date",
    "event_uses_venue",
    "event_uses_space",
    "event_date_uses_venue",
    "event_date_uses_space",
    "user_member_of_organization",
    "user_invited_to_organization",
    "organization_partner_request",
    "organization_partner_of",
]


class GraphRoot(BaseModel):
    type: GraphEntityType
    key: UUID


class GraphFilters(BaseModel):
    model_config = ConfigDict(extra="forbid")

    root_type: GraphEntityType
    root_key: UUID
    depth: int = Field(default=2, ge=1, le=3)
    relation_type: GraphRelationType | None = None


class GraphSearchFilters(BaseModel):
    geo_scope_id: UUID | None = None
    model_config = ConfigDict(extra="forbid")

    q: str = Field(min_length=2, max_length=120)

    @field_validator("q", mode="before")
    @classmethod
    def trim_query(cls, value: str) -> str:
        return value.strip()

    entity_type: GraphEntityType | None = None
    organization_id: UUID | None = None
    limit: int = Field(default=20, ge=1, le=20)


class GraphNode(BaseModel):
    id: str
    type: GraphEntityType
    key: UUID
    label: str
    subtitle: str | None = None
    status: str | None = None
    public_url: str | None = None
    admin_url: str | None = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    type: GraphRelationType
    label: str
    direction: Literal["directed", "undirected"] = "directed"


class GraphResponse(BaseModel):
    root: GraphRoot
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    truncated: bool
    max_nodes: int = 100
    max_edges: int = 200


class GraphSearchResponse(BaseModel):
    items: list[GraphNode]

File: frontend/tests/fixtures/graph.ts
import type {
  GraphNode,
  GraphEdge,
  GraphEntityType,
  GraphRelationType,
  GraphResponse,
} from '../../shared/contracts'
const node = (type: GraphEntityType, number: number, label: string): GraphNode => {
  const key = `20000000-0000-4000-8000-${String(number).padStart(12, '0')}`
  const sections = {
    organization: 'organizations',
    venue: 'venues',
    space: 'spaces',
    event: 'events',
    user: 'users',
  }
  return {
    id: `${type}:${key}`,
    type,
    venue_scope: type === 'venue' ? (number === 4 ? 'shared' : 'organization') : null,
    key,
    label,
    subtitle: null,
    status: null,
    public_url: null,
    admin_url:
      type === 'event_date'
        ? `/activity?entity_key=${key}&entity_type=${type}`
        : `/${sections[type]}/${key}`,
  }
}
const nodes = [
  node('organization', 1, 'Kulturzentrum Rendsburg e.V.'),
  node('user', 2, 'Max Mustermann'),
  node('user', 3, 'Anna Beispiel'),
  node('venue', 4, 'Kulturzentrum Rendsburg'),
  node('venue', 5, 'St. Marien-Kirche Rendsburg'),
  node('space', 6, 'Großer Saal'),
  node('space', 7, 'Proberaum'),
  node('space', 8, 'Kirchensaal'),
  node('event', 9, 'Lange Nacht der Kirchen'),
  node('event', 10, 'Jazz im Hof'),
  node('event_date', 11, 'Lange Nacht · 20.09.2026'),
  node('event_date', 12, 'Jazz · 20.09.2026'),
]
const edge = (a: number, b: number, type: GraphRelationType, label: string): GraphEdge => ({
  id: `${type}:${a}:${b}`,
  source: nodes[a]!.id,
  target: nodes[b]!.id,
  type,
  label,
  direction: 'directed',
})
export const graphFixture: GraphResponse = {
  root: { type: 'organization', key: nodes[0]!.key },
  nodes,
  edges: [
    edge(1, 0, 'user_member_of_organization', 'Mitglied von'),
    edge(2, 0, 'user_member_of_organization', 'Mitglied von'),
    edge(0, 3, 'organization_has_venue', 'Betreibt'),
    edge(0, 4, 'organization_has_venue', 'Betreibt'),
    edge(3, 5, 'venue_has_space', 'Hat Raum'),
    edge(3, 6, 'venue_has_space', 'Hat Raum'),
    edge(4, 7, 'venue_has_space', 'Hat Raum'),
    edge(0, 8, 'organization_has_event', 'Organisiert'),
    edge(0, 9, 'organization_has_event', 'Organisiert'),
    edge(8, 4, 'event_uses_venue', 'Standardort'),
    edge(8, 10, 'event_has_date', 'Hat Termin'),
    edge(9, 11, 'event_has_date', 'Hat Termin'),
  ],
  truncated: false,
  max_nodes: 100,
  max_edges: 200,
}
export const graphPath = `/graph?root_type=organization&root_key=${nodes[0]!.key}&depth=2`
